core/model/src/main/kotlin/au/com/dius/pact/core/model/matchingrules/MatchingRules.kt

```diff
@@ -134,6 +134,38 @@ object ValuesMatcher : MatchingRule {
   override fun toMap(spec: PactSpecVersion) = toMap()
 }
 
+/**
+ * Matcher for ignoring order of elements in array
+ */
+object IgnoreOrderMatcher : MatchingRule {
+  override fun toMap() = mapOf("match" to "ignore-order")
+  override fun toMap(spec: PactSpecVersion) = toMap()
+}
+
+/**
+ * Ignore order matching with a maximum size
+ */
+data class MaxEqualsIgnoreOrderMatcher(val max: Int) : MatchingRule {
```

Comment:
I'm not sure why you mixed in equality into the Min/MaxIgnoreOrderMatchers?  Ignore order matching, regardless of whether or not it is combined with min or max, isn't unique to equality matching.  It should also work when combined with things like regex matching.

```suggestion
data class MaxIgnoreOrderMatcher(val max: Int) : MatchingRule {
```

Reply:
As I understand it, the Min/Max/MinMaxTypeMatcher is only good for matching the size constraints *and* Type.  This is the reason why I couldn't wrap my head around why we would want to mix it with IgnoreOrderMatcher.  Since type doesn't care about the order.

I explicitly named them with Equals because they wouldn't work with some other type of matcher such as DateMatcher for example.

I don't know where that leaves us regarding the name change or perhaps the implementation of Min/Max/MinMax ignore order matchers.

Reply:
But ignoreorder matchers (with or without the min/max) can be combined with any other type of matching, including type matching.

An example of this is in [`'matching json bodies - with ignore-order - and mixed elements'`](https://github.com/mitre/pact-jvm/blob/a57de008f7a54d7d4ea78ca39d9da763854a6a39/core/matchers/src/test/groovy/au/com/dius/pact/core/matchers/JsonBodyMatcherSpec.groovy#L557) (the below snippet with the changes I had suggested previously).  Here, we want to do type matching on `4` in expected, but don't care in what position in the array a number shows up (and want to this not to match if a number isn't present):
```groovy
  @Unroll
  def 'matching json bodies - with ignore-order - and mixed elements'() {
    given:
    matchers.addCategory('body')
            .addRule('$', IgnoreOrderMatcher.INSTANCE)
            .addRule('$[*].id', new EqualsMatcher())
            .addRule('$[0].status', new EqualsMatcher())
            .addRule('$[1]', TypeMatcher.INSTANCE)
            .addRule('$[2].status', new RegexMatcher('up|down'))
    def actualBody = OptionalBody.body(actual.bytes)
    def expectedBody = OptionalBody.body('[{"id":"a", "status":"up"}, 4, {"id":"b", "status":"down"}]'.bytes)

    expect:
    matcher.matchBody(expectedBody, actualBody, true, matchers).empty == matches

    where:
    actual                                                          | matches
    '[{"id":"a", "status":"up"}, 4, {"id":"b", "status":"down"}]'   | true
    '[{"id":"a", "status":"up"}, "4", {"id":"b", "status":"down"}]' | false
    '[{"id":"b", "status":"up"}, {"id":"a", "status":"up"}, 5]'     | true
    '[{"id":"b", "status":"down"}, {"id":"a", "status":"up"}, 5]'   | true
    '[{"id":"b", "status":"up"}, {"id":"a", "status":"down"}, 5]'   | false
    '[{"id":"c", "status":"up"}, {"id":"a", "status":"up"}, 5]'     | false
  }
```

I had also suggested this other test case, which also provides an example of how this could work, where you still want to make sure the expected number is in the actual list (with an equality matcher, ignoring order) but also want to make sure that all the other items in the list are numbers (with a `$[*]` type matcher):
```groovy
  def 'matching json bodies - return type mismatch - with ignore-order and min-type matching'() {
    given:
    matchers.addCategory('body')
            .addRule('$', new MinEqualsIgnoreOrderMatcher(1))
            .addRule('$[0]', EqualsMatcher.INSTANCE)
            .addRule('$[*]', TypeMatcher.INSTANCE)
    when:
    def mismatches = matcher.matchBody(expectedBody, actualBody, true, matchers)
    then:
    !mismatches.empty
    mismatches*.mismatch == ['Expected "bad" (JsonPrimitive) to be the same type as 100 (JsonPrimitive)']
    mismatches*.path == ['$.3']
    where:
    actualBody = OptionalBody.body('[200, 100, 300, "bad"]'.bytes)
    expectedBody = OptionalBody.body('[100]'.bytes)
  }
```

The only special relationship I see between ignoreorder and equality is that ignoreorder matching should fall-back to basic equality matching if you apply that rule to a non-list, but it does this mostly as equality matching is the default form of matching in Pact.  But I'm not sure if that justifies it being in the name when weighed agains the fact that ignore-order matchers can be paired with other matchers...  If we do keep the Equals in the name here, we should also change `IgnoreOrderMatcher` to `EqualsIgnoreOrderMatcher` for consistency.

Reply:
It all started with the test below.  However, we were trying to combine an IgnoreOrderMatcher with a MinTypeMatcher.  But since MinTypeMatcher only matches Types, all of the tests passed - even the third one which was supposed to fail.

There may have been another way to combine matchers and get the correct behavior.  I don't know.  However, we discussed this other MinIgnoreOrderMatcher - which became MinEqualsIgnoreOrderMatcher because the min/max/minmax matchers have another dimension to check against beyond just the array length (i.e. the individual elements).

I'll keep at this and think about it more.  We may want to setup a meeting to discuss further. 

```
@Unroll
  def 'matching json bodies - with ignore-order and min-equals matching - when actual has extra elements'() {
    given:
    def actualBody = OptionalBody.body(actual.bytes)
    def expectedBody = OptionalBody.body(expected.bytes)
    matchers.addCategory('body')
        .addRule('$', new MinEqualsIgnoreOrderMatcher(3))

    when:
    def mismatches = matcher.matchBody(expectedBody, actualBody, true, matchers)

    then:
    mismatches.empty == match

    where:
    expected                | actual                            | match
    '[1,2]'                 | '[1,2,3]'                         | true
    '[1,2]'                 | '[2,1,3]'                         | true
    '[1,2]'                 | '[1,3,4]'                         | false
    '[{"i":"a"},{"i":"b"}]' | '[{"i":"a"},{"i":"b"},{"i":"c"}]' | true
  }

```

Reply:
I've been thinking that it may be possible to pass the intended element matcher into Min/Max/MinMax IgnoreOrderMatcher along with the size(s).  Then rather than defaulting to Type or Equals, or whatever it could just use the matcher provided.

Reply:
The `.addRule('$', new MinEqualsIgnoreOrderMatcher(3))` in that test was meant to replace the following [four rules](https://github.com/mitre/pact-jvm/blob/f37d290c63a3069196980b0bc91c296564e4045e/core/matchers/src/test/groovy/au/com/dius/pact/core/matchers/JsonBodyMatcherSpec.groovy#L429):
```groovy
        .addRule('$', IgnoreOrderMatcher.INSTANCE)
        .addRule('$[0]', EqualsMatcher.INSTANCE)  // explicit EqualsMatcher needed to override MinTypeMatcher
        .addRule('$[1]', EqualsMatcher.INSTANCE)
        .addRule('$', new MinTypeMatcher(3))
```
In the original 4-rule version the MinTypeMatcher was used on the array itself to ensure that at least 3 elements were present in actual, but type matching shouldn't be applied to the elements because of the two explicit EqualsMatcher which override the type matching for those elements.  The IgnoreOrderMatcher allowed the 2 expected element values to be located out of order in the actual array.

Here is another version of the test with the ignore-order aspects commented out, for a point of reference. This all passes, including the test case with a false match.
```groovy
  @Unroll
  def 'matching json bodies - with ignore-order and min-type matching - when actual has extra elements'() {
    given:
    def actualBody = OptionalBody.body(actual.bytes)
    def expectedBody = OptionalBody.body(expected.bytes)
    matchers.addCategory('body')
            //.addRule('$', IgnoreOrderMatcher.INSTANCE)
            .addRule('$[0]', EqualsMatcher.INSTANCE)  // explicit EqualsMatcher needed to override MinTypeMatcher
            .addRule('$[1]', EqualsMatcher.INSTANCE)
            .addRule('$', new MinTypeMatcher(3))

    expect:
    matcher.matchBody(expectedBody, actualBody, true, matchers).empty == match

    where:
    expected                | actual                            | match
    '[1,2]'                 | '[1,2,3]'                         | true
    //'[1,2]'                 | '[2,1,3]'                         | true
    '[1,2]'                 | '[1,3,4]'                         | false
    '[{"i":"a"},{"i":"b"}]' | '[{"i":"a"},{"i":"b"},{"i":"c"}]' | true
  }
```

The intent of MinIgnoreOrderMatcher was to just combine the Min feature of MinTypeMatcher with the IgnoreOrderMatcher, so that we could specify the need for an array which could be larger than the expected example without using a MinTypeMatcher and then overriding the type matching on the elements of the array.

I'm not sure I follow the need to pass another matcher into a Min/Max/MinMaxIgnoreOrderMatcher.  I think the current behavior of falling back to an equals matcher is correct (I had originally thought you should fallback to `matchIgnoringOrder` but later realized that `matchIgnoringOrder` should also just result in `matchEquals` in that situation).  If an element of the array needs to do something other than equals, you just define a matcher for that element (e.g. `$[0]`) or all elements (i.e. `$[*]`).